README.md
# Introduction
[![NPM Version][npm-badge]][npm-url]
[![NPM Downloads][npmd-badge]][npm-url]
[![Build Status][travis-badge]][travis-url]
[![Test Coverage][codecov-badge]][codecov-url]
[![Dependencies][dependencies-badge]][dependencies-url]
[![devDependencies][devDependencies-badge]][devDependencies-url]
[![Known Vulnerabilities][snyk-badge]][snyk-url]

A simple logging middleware for the [koa] http framework for nodejs. This module uses the [pino] logger and was inspired by the [koa-pino-logger] module. This module was created with typescript, and is compatible with both javascript and typescript projects.

As well as logging requests and providing a log object in requests, this module also sets the HTTP Headers Date, X-Response-Time and X-Request-ID.
- **X-Request-ID** Header is set to a new uuid or the value of the X-Request-ID header sent with the request. This allows requests to be tracked through microservices.
- **Date** Header is set to the date and time that the request was received.
- **X-Response-Time** Header is set to the response time of the request in milliseconds.

## Contents
- [Install](#Install)
- [Usage](#Usage)
- [Test](#Test)
- [License](#License)

# Install
```
yarn add koa-req-logger
npm install koa-req-logger
```

# Usage

For a full API Reference see the documentation [here⇗](https://drbarnabus.github.io/koa-req-logger/globals.html).

### TypeScript
```ts
import Koa from 'koa';
import { KoaReqLogger, KoaReqLoggerOptions } from 'koa-req-logger';

const opts: KoaReqLoggerOptions = {
  disableIdHeader: false;
};

const app = new Koa();
const logger = new KoaReqLogger(opts);
app.use(logger.getMiddleware());

app.use((ctx, next) => {
  ctx.log.info('Some Log Message');
  ctx.log.warn({ obj: 'object' }, 'Log a message with an object');

  ctx.throw(400, 'Bad Request');
});

app.listen();
```

### JavaScript
```js
const Koa = require('koa');
const { KoaReqLogger } = require('koa-req-logger');

const app = new Koa();

const logger = new KoaReqLogger();
app.use(logger.getMiddleware());

app.use((ctx, next) => {
  ctx.log.info('Some Log Message');
  ctx.log.warn({ obj: 'object' }, 'Log a message with an object');

  ctx.throw(400, 'Bad Request');
});

app.listen(3000);
```

### Output
Produces a similar output to the following json, which can then be parsed with pino's shell utility to pretty-print the output.

```json
{"level":30,"time":1532251116578,"msg":"::1 - GET /","pid":4992,"hostname":"server.local","id":"ff0bae4b-b067-4cd6-8b99-5d221e74c515","req":{"method":"GET","url":"/","headers":{"host":"localhost:3000","connection":"keep-alive","upgrade-insecure-requests":"1","user-agent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/67.0.3396.99 Safari/537.36","accept":"text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8","accept-encoding":"gzip, deflate, br","accept-language":"en-GB,en-US;q=0.9,en;q=0.8"}},"startDate":"Sun, 22 Jul 2018 09:18:36 GMT","v":1}
{"level":30,"time":1532251116579,"msg":"Some Log Message","pid":4992,"hostname":"server.local","id":"ff0bae4b-b067-4cd6-8b99-5d221e74c515","v":1}
{"level":40,"time":1532251116579,"msg":"Log a message with an object","pid":4992,"hostname":"server.local","id":"ff0bae4b-b067-4cd6-8b99-5d221e74c515","obj":"object","v":1}
{"level":50,"time":1532251116583,"msg":"::1 - GET / - 400 4ms","pid":4992,"hostname":"server.local","id":"ff0bae4b-b067-4cd6-8b99-5d221e74c515","res":{"status":400,"headers":{"x-request-id":"ff0bae4b-b067-4cd6-8b99-5d221e74c515","date":"Sun, 22 Jul 2018 09:18:36 GMT","x-response-time":"4ms","content-type":"application/json; charset=utf-8"}},"err":{"type":"ClientError","message":"Bad Request","stack":"BadRequestError: Bad Request\n    at Object.throw (/home/drbarnabus/Development/test-service/node_modules/koa/lib/context.js:96:11)...","status":400,"statusCode":400,"expose":true},"responseTime":4,"startDate":"Sun, 22 Jul 2018 09:18:36 GMT","v":1}
```

# Test
```
yarn test
npm test
```

# License
Licensed under [MIT](./LICENSE).

<!-- Links --->
[koa]: https://github.com/koajs/koa
[pino]: https://github.com/pinojs/pino
[koa-pino-logger]: https://github.com/pinojs/koa-pino-logger
[koa-router]: https://github.com/alexmingoia/koa-router

<!-- Badges --->
[npm-badge]: https://img.shields.io/npm/v/koa-req-logger.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/koa-req-logger
[npmd-badge]: https://img.shields.io/npm/dw/koa-req-logger.svg?style=flat-square
[travis-badge]: https://img.shields.io/travis/DrBarnabus/koa-req-logger/master.svg?style=flat-square
[travis-url]: https://travis-ci.org/DrBarnabus/koa-req-logger
[dependencies-badge]: https://david-dm.org/drbarnabus/koa-req-logger.svg?style=flat-square
[codecov-badge]: https://img.shields.io/codecov/c/github/DrBarnabus/koa-req-logger/master.svg?style=flat-square
[codecov-url]: https://codecov.io/gh/DrBarnabus/koa-req-logger
[dependencies-url]: https://david-dm.org/drbarnabus/koa-req-logger
[devDependencies-badge]: https://david-dm.org/drbarnabus/koa-req-logger/dev-status.svg?style=flat-square
[devDependencies-url]: https://david-dm.org/drbarnabus/koa-req-logger?type=dev
[snyk-badge]: https://snyk.io/test/github/DrBarnabus/koa-req-logger/badge.svg?targetFile=package.json&style=flat-square
[snyk-url]: https://snyk.io/test/github/DrBarnabus/koa-req-logger?targetFile=package.json

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file. This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.5.3] - 2019-08-02
### Changed
- Updated package dependencies.

## [1.5.1] - 2018-09-20
### Fixed
- Rebuilt the package and republished as the previous version didn't build correctly.

## [1.5.0] - 2018-09-19
### Changed
- **Breaking Change:** To support user supplied instances of Pino and extreme mode, the way that options are passed to pino has been changed significantly. Please refer to the new documentation to see how to update your implementation of KoaReqLogger.

## [1.4.0] - 2018-09-04
### Changed
- Changed the options to disable the headers again, the conversion to TypeScript left them reversed so the option enabled it as opposed to disabling. The options are now called disableIdHeader, disableDateHeader and disableResponseTimeHeader. If any set to true, the corresponding header will be disabled.
### Fixed
- Fixed an issue where the header disable options were reversed due to the TypeScript conversion.
- FIxed a typo in readme documentation.

## [1.3.8] - 2018-09-03
### Changed
- Updated README to reflect import for JavaScript.

## [1.3.3] - 2018-09-03
### Fixed
- Fixed default export so that the module works correctly in JavaScript projects.

## [1.3.2] - 2018-09-03
### Fixed
- Added default export so that the module works correctly in JavaScript projects.

## [1.3.1] - 2018-09-03
### Fixed
- Fixed npmignore to include build JavaScript files.

## [1.3.0] - 2018-09-03
### Added
- Added a TypeScript type declaration file so that the module can be more easily used in a typescript project.
### Changed
- Updated package dependencies.
- Updated README.md with examples of usage in TypeScript.
- Updated API Reference documentation.
- **Breaking Change** to disable specific headers or all headers, you are now required to set idHeader, dateHeader and responseTimeHeader respectively. To disable all headers all three options must be set.

## [1.2.2] - 2018-08-08
### Changed
- Updated package dependencies to use the latest version of pino v5.0.1 as a minimum.

## [1.2.1] - 2018-08-01
### Fixed
- Fixed an issue where the logger didn't correctly handle errors thrown without a status. These are now handled as status 500 internal server errors to prevent it not being logged. An example would be <code>throw new Error('Generic Error with no Status');</code> which will now respond with a generic status 500 error.

## [1.2.0] - 2018-07-23
### Added
- Added functionality to allow each individual header to be turned off. This can be configured by setting one of; headers.id, headers.date and headers.responseTime to false in the options object passed to KoaReqLogger.
- Added functionality to allow all headers to be disabled by setting headers to false in the options object passed to KoaReqLogger.
- Added tests to get code coverage up to 100%.
- Added API Reference file, documents the options that are available when creating a new instance of KoaReqLogger.
- Added ci to test builds, badges also added to readme to show build status.
### Changed
- By default, errors are now logged as warnings. If the error status is 5** then the error is logged as an error. But this can be changed back to the original behavior, by setting alwaysError to true in the options object passed to KoaReqLogger.
### Fixed
- Fixed a problem in the readme file example where the wrong class name was used.

## [1.1.0] - 2018-07-21
### Added
- Functionality to override the uuidv4 that is used automatically for the X-Request-ID with your own uuid function.
- Updated the readme file, with an example of log usage in a request.
### Fixed
- Fixed an issue where response object wasn't populated correctly in the log.

## 1.0.0 - 2018-07-20
First release of the package.
### Added
- Added this CHANGELOG file to keep track of changes.
- Implemented the middleware with configureable serializers standard are inserted automatically. Remaining options are just passed to pino without any validation at present.
- Implemented basic tests to ensure that the 3 HTTP Headers; Date, X-Request-ID and X-Response-Time are added to the response correctly.
- Documented the fix for an issue that was encounted when using the koa-router package.



<!-- LINKS --->
[1.5.3]: https://github.com/DrBarnabus/koa-req-logger/compare/v1.5.2...v1.5.3
[1.5.1]: https://github.com/DrBarnabus/koa-req-logger/compare/v1.5.0...v1.5.1
[1.5.0]: https://github.com/DrBarnabus/koa-req-logger/compare/v1.4.0...v1.5.0
[1.4.0]: https://github.com/DrBarnabus/koa-req-logger/compare/v1.3.8...v1.4.0
[1.3.8]: https://github.com/DrBarnabus/koa-req-logger/compare/v1.3.3...v1.3.8
[1.3.3]: https://github.com/DrBarnabus/koa-req-logger/compare/v1.3.2...v1.3.3
[1.3.2]: https://github.com/DrBarnabus/koa-req-logger/compare/v1.3.1...v1.3.2
[1.3.1]: https://github.com/DrBarnabus/koa-req-logger/compare/v1.3.0...v1.3.1
[1.3.0]: https://github.com/DrBarnabus/koa-req-logger/compare/v1.2.2...v1.3.0
[1.2.2]: https://github.com/DrBarnabus/koa-req-logger/compare/v1.2.1...v1.2.2
[1.2.1]: https://github.com/DrBarnabus/koa-req-logger/compare/v1.2.0...v1.2.1
[1.2.0]: https://github.com/DrBarnabus/koa-req-logger/compare/v1.1.0...v1.2.0
[1.1.0]: https://github.com/DrBarnabus/koa-req-logger/compare/v1.0.0...v1.1.0

index.js
const pino = require('pino');
const uuidv4 = require('uuid/v4');

// Set of standard serializers
const stdSerializers = {
  req: require('./lib/req'),
  res: require('./lib/res'),
  err: require('./lib/err')
};

module.exports = class KoaReqLogger {
  /**
   * A logging middleware for koa, this middleware also sets the HTTP Date header,
   * sets the X-Response-Time header to the response time in milliseconds and sets the
   * X-Request-ID header to either a new uuidv4 or the request id passed into the request
   * @example
   * ``` javascript
   * const app = Koa();
   * const logger = KoaReqLogger();
   * app.use(logger.getMiddleware());
   * app.listen(3000);
   * ```
   * @param {object} opts An optional options object
   * @api public
   */
  constructor(opts) {
    this.middleware = this.middleware.bind(this);

    opts = opts || {};

    // Set standard serializers, if no custom ones used
    opts.serializers = opts.serializers || {};
    opts.serializers.req = opts.serializers.req || stdSerializers.req;
    opts.serializers.res = opts.serializers.res || stdSerializers.res;
    opts.serializers.err = opts.serializers.err || stdSerializers.err;

    // Check if all headers should be disabled
    if (typeof opts.headers === 'boolean') {
      if (opts.headers == true) {
        this.idHeader = true;
        this.startHeader = true;
        this.responseTimeHeader = true;
      } else {
        this.idHeader = false;
        this.startHeader = false;
        this.responseTimeHeader = false;
      }
    } else {
      opts.headers = opts.headers || {};

      // Check if X-Request-ID Header should be disabled
      if (opts.headers.id !== undefined) {
        this.idHeader = opts.headers.id;
      } else {
        this.idHeader = true;
      }

      // Check if Date Header should be disabled
      if (opts.headers.date !== undefined) {
        this.startHeader = opts.headers.date;
      } else {
        this.startHeader = true;
      }

      // Check if X-Response-Time should be disabled
      if (opts.headers.responseTime !== undefined) {
        this.responseTimeHeader = opts.headers.responseTime;
      } else {
        this.responseTimeHeader = true;
      }
    }

    delete opts.headers;

    // Check if a uuidFunction has been passed in options and use if available
    if (typeof opts.uuidFunction === 'function') {
      this.uuidFunction = opts.uuidFunction;
      delete opts.uuidFunction;
    } else {
      this.uuidFunction = uuidv4;
    }

    delete opts.uuidFunction;

    this.alwaysError = opts.alwaysError || false;

    delete opts.alwaysError;

    this.logger = pino(opts);
  }

  /**
   * This function takes a request context and assigns a request id
   * This will either be a newly generated uuidv4 or the X-Request-ID header passed into the request
   * The resulting request id is then set as the X-Request-ID header on the response
   * @param ctx The current koa response context
   * @api private
   */
  setRequestId(ctx) {
    if (ctx.get('X-Request-ID')) {
      ctx.id = ctx.get('X-Request-ID');
    } else {
      ctx.id = this.uuidFunction();
    }

    if (this.idHeader) {
      ctx.set('X-Request-ID', ctx.id);
    }
  }

  /**
   * This function sets the current date as the date header,
   * as well as logging the request and storing the start date for the calculation of the response time later
   * @param ctx The current koa context
   * @api private
   */
  startRequest(ctx) {
    ctx.start = new Date();

    if (this.startHeader) {
      ctx.set('Date', ctx.start.toUTCString());
    } else {
      ctx.remove('Date'); // Remove default header set by koa
    }

    ctx.log.info({ req: ctx.req, startDate: ctx.start.toUTCString() }, `${ctx.request.ip} - ${ctx.method} ${ctx.path}`);
  }

  /**
   * This function calcuates the response time and sets the X-Response-Time header
   * @param ctx The current koa context
   * @api private
   */
  setResponseTime(ctx) {
    let now = new Date();
    ctx.responseTime = now - ctx.start;

    if (this.responseTimeHeader) {
      ctx.set('X-Response-Time', `${ctx.responseTime}ms`);
    }
  }

  /**
   * This function logs the end of a request,
   * it also calls another function to calculate and set the X-Response-Time header
   * @param ctx The current koa context
   * @api private
   */
  endRequest(ctx) {
    this.setResponseTime(ctx);

    ctx.log.info(
      { res: ctx.response, responseTime: ctx.responseTime, startDate: ctx.start.toUTCString() },
      `${ctx.ip} - ${ctx.method} ${ctx.path} - ${ctx.status} ${ctx.responseTime}ms`
    );
  }

  /**
   * This function logs the end of a request and the error that has been caught,
   * it also calls another function to calculate and set the X-Response-Time header
   * @param ctx The current koa context
   * @api private
   */
  endRequestError(e, ctx) {
    this.setResponseTime(ctx);

    // Construct error response
    if (e.status !== undefined) {
      ctx.status = e.status;
      ctx.body = {
        error: {
          code: e.status,
          message: e.message
        }
      };
    } else {
      ctx.status = 500;
      ctx.body = {
        error: {
          code: 500,
          message: 'Internal Server Error'
        }
      };
    }

    if (((ctx.status / 100) | 0) == 5 || this.alwaysError) {
      ctx.log.error(
        { res: ctx.response, err: e, responseTime: ctx.responseTime, startDate: ctx.start.toUTCString() },
        `${ctx.ip} - ${ctx.method} ${ctx.path} - ${ctx.status} ${ctx.responseTime}ms`
      );
    } else {
      ctx.log.warn(
        { res: ctx.response, err: e, responseTime: ctx.responseTime, startDate: ctx.start.toUTCString() },
        `${ctx.ip} - ${ctx.method} ${ctx.path} - ${ctx.status} ${ctx.responseTime}ms`
      );
    }
  }

  /**
   * This function handles all requests and is used as a koa middleware
   * @param ctx The current koa context
   * @param next The next function in the middleware stack
   * @api private
   */
  async middleware(ctx, next) {
    // Set the request id
    this.setRequestId(ctx);

    // Create a child of the logger with the request id as the key
    ctx.log = ctx.logger = ctx.req.log = ctx.request.log = ctx.res.log = this.logger.child({ id: ctx.id });

    // Start the request logging
    this.startRequest(ctx);

    try {
      await next();

      // End the request logging
      this.endRequest(ctx);
    } catch (e) {
      // Handle error and end request logging
      this.endRequestError(e, ctx);
    }
  }

  /**
   * This function returns the middleware function for use in koa
   * @api public
   */
  getMiddleware() {
    return this.middleware;
  }
};
